### PR TITLE
feat: make tests auto generate from file

### PR DESCRIPTION
**Pull request by KolbyML**
I also made it so tests will be auto generated from a test file of content.

Order of the test file matters so it should be

Header
(optional bb or receipt)
(optional bb or receipt)
The implementation should be flexibile enough so different schemes can be done but this is the scheme for history content.

### CODE REVIEW

File: simulators/portal-interop/src/main.rs [hunk 1]
```diff
@@ -237,574 +209,104 @@ dyn_async! {
...
-        let header_key: HistoryContentKey =
-            serde_yaml::from_value(values[10].get("content_key").unwrap().clone()).unwrap();
-        let header_value: HistoryContentValue =
-            serde_yaml::from_value(values[10].get("content_value").unwrap().clone()).unwrap();
-        let content_key: HistoryContentKey =
-            serde_yaml::from_value(values[12].get("content_key").unwrap().clone()).unwrap();
-        let content_value: HistoryContentValue =
-            serde_yaml::from_value(values[12].get("content_value").unwrap().clone()).unwrap();
-        test_offer_x(client_a, client_b, (content_key, content_value), Some((header_key, header_value))).await;
+        match result {
```
Comment (njgheorghita) [inline simulators/portal-interop/src/main.rs:270]:
```rust
if let Ok(ContentInfo::Enrs { enrs: val }) = result {
    if !val.is_empty() {
        panic!("Error: Unexpected FINDCONTENT response: expected ContentInfo....");
    }
} else {
    panic!("Error....");
}
```

warning: I didn't test the above, but I think this might help get rid of some of these lines, but only if you don't think it's important to have all of these different error messages for debugging

File: simulators/portal-interop/src/main.rs [hunk 2]
```diff
@@ -237,574 +209,104 @@ dyn_async! {
...
-                    client_b: client_b.clone(),
-                }
-            ).await;
+            for ProcessedContent { content_type, block_number, test_data } in process_content(content.clone()) {
+                test.run(
+                    NClientTestSpec {
+                        name: format!("RecursiveFindContent {}: block number {}{} {} --> {}", content_type, block_number, get_flair(block_number), client_a.name, client_b.name),
+                        description: "".to_string(),
+                        always_run: false,
+                        run: test_recursive_find_content_x,
```
Comment (njgheorghita) [inline simulators/portal-interop/src/main.rs:218]: Also here, I'd be extremely happy if we could just rename all of the `_x` suffix fns. Imo the `_x` suffix really isn't adding anything useful/descriptive in the fn name, and simply dropping it won't make it more difficult to understand what's going on

File: simulators/portal-interop/src/main.rs [hunk 3]
```diff
@@ -54,155 +57,124 @@ async fn run_suite(host: Simulation, suite: Suite) {
...
+        "".to_string()
+    }
+}
+
 dyn_async! {
    async fn test_portal_interop<'a> (test: &'a mut Test, _client: Option<Client>) {
         // Get all available portal clients
         let clients = test.sim.client_types().await;
 
+        let values = std::fs::read_to_string("./test-data/test_data_collection_of_forks_blocks.yaml")
```
Comment (njgheorghita) [inline simulators/portal-interop/src/main.rs:152]: Could you move this filename to a const w/ a brief comment describing what kind of data should go in the file / why one might add more data to the file / what tests are run for any data in the file

**Review by njgheorghita — COMMENTED** [review]
Honestly, this is very nice and refreshing to see such a big negative diff and getting rid of some very un-[DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) code!

File: simulators/portal-interop/src/main.rs [hunk 4]
```diff
@@ -237,574 +209,104 @@ dyn_async! {
...
+                panic!("Error: Unable to get response from FINDCONTENT request: {err:?}");
+            }
+        }
     }
 }
 
 dyn_async! {
-    async fn test_offer_x<'a>(client_a: Client, client_b: Client, target_content: (HistoryContentKey, HistoryContentValue), optional_content: Option<(HistoryContentKey, HistoryContentValue)>) {
-        if let Some((optional_key, optional_value)) = optional_content {
+    async fn test_offer_x<'a>(clients: Vec<Client>, test_data: Option<Vec<(String, String)>>) {
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply simulators/portal-interop/src/main.rs:294]: resolved

File: simulators/portal-interop/src/main.rs [hunk 5]
```diff
@@ -237,574 +209,104 @@ dyn_async! {
...
+        let (client_a, client_b) = match clients.iter().collect_tuple() {
+            Some((client_a, client_b)) => (client_a, client_b),
+            None => {
+                panic!("Unable to get expected amount of clients from NClientTestSpec");
+            }
+        };
+        let test_data = match test_data {
+            Some(test_data) => test_data,
+            None => panic!("Expected test data non was provided"),
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply simulators/portal-interop/src/main.rs:303]: resolved

File: simulators/portal-interop/src/main.rs [hunk 6]
```diff
@@ -54,155 +57,124 @@ async fn run_suite(host: Simulation, suite: Suite) {
...
+            content_type,
+            block_number,
+            test_data,
+        })
+    }
+    result
+}
+
+fn get_flair(block_number: u64) -> String {
+    if block_number > 17034870 {
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply simulators/portal-interop/src/main.rs:126]: https://ethereum.org/en/history/

Sure I will do constants

[the remark this answers is not in the document]
Comment (KolbyML) [reply simulators/portal-interop/src/main.rs:126]: resolved

File: simulators/portal-interop/src/main.rs [hunk 7]
```diff
@@ -237,574 +209,104 @@ dyn_async! {
...
+                        test_data: Some(test_data),
+                        clients: vec![client_a.clone(), client_b.clone()],
+                    }
+                ).await;
+            }
 
             // Test gossiping a collection of blocks to node B (B will gossip back to A)
             test.run(
-                TwoClientTestSpec {
+                NClientTestSpec {
```
[the remark this answers is not in the document]
Comment (KolbyML) [reply simulators/portal-interop/src/main.rs:242]: resolved

File: simulators/portal-interop/src/main.rs (hunk 1 above)
Comment (KolbyML) [reply simulators/portal-interop/src/main.rs:270]: I think it would be helpful if someone was debugging it so I will leave it now it is

File: simulators/portal-interop/src/main.rs (hunk 3 above)
Comment (KolbyML) [reply simulators/portal-interop/src/main.rs]: resolved

**Comment by KolbyML** [comment]
@njgheorghita ok I believe I resolved your PR concerns so it should be ready for another review now

File: simulators/portal-interop/src/constants.rs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+// block receipt 2
+// block header 3
+// would be a valid test but
+// block header 2
+// block receipt 3
+// block header 3
+// this wouldn't be a valid test file as block receipt 3 doesn't have a header place before it
+// related header, block body, receipts must be sequential
+// We want this so block forks with different formats can be tested, by updating an external test repo
+pub const TEST_DATA_FILE_PATH: &str = "./test-data/test_data_collection_of_forks_blocks.yaml";
```
Comment (njgheorghita) [inline simulators/portal-interop/src/constants.rs:29]: I'm guessing we will eventually add more test data to run these tests for other subnetworks. In which case, I'd imagine we'll have a separate test-file with that data (since it could need special treatment in how it's gossiped). Basically, I think a filename of `./test-data/history.yaml` is sufficient here. The current filename is kinda confusing. 

I like the improved description for how the test file should be configured, but I think that it would be better off at the top of the test file itself, since it's not obvious that future devs will see this page when they're adding tests.

I'd also add a comment listing exactly what tests will be generated & run on each content key/value pair that's added to the test data

File: simulators/portal-interop/src/constants.rs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+pub const BERLIN_BLOCK_NUMBER: u64 = 12244000;
+pub const ISTANBUL_BLOCK_NUMBER: u64 = 9069000;
+pub const CONSTANTINOPLE_BLOCK_NUMBER: u64 = 7280000;
+pub const BYZANTIUM_BLOCK_NUMBER: u64 = 4370000;
+pub const HOMESTEAD_BLOCK_NUMBER: u64 = 1150000;
+
+// Tests can be written to be generate from this file.
+// Tests data will be in a yaml key value format
+// Tests consist of content key value pairs.
+// For history tests the header for the respective block body/receipt must be place before
```
Comment (njgheorghita) [inline simulators/portal-interop/src/constants.rs:14]:
```suggestion
// For history tests, the header MUST be placed before its respective block body/receipt, to ensure that the data necessary for validation is available.
```

File: simulators/portal-interop/src/constants.rs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+pub const MERGE_BLOCK_NUMBER: u64 = 15537394;
+pub const LONDON_BLOCK_NUMBER: u64 = 12965000;
+pub const BERLIN_BLOCK_NUMBER: u64 = 12244000;
+pub const ISTANBUL_BLOCK_NUMBER: u64 = 9069000;
+pub const CONSTANTINOPLE_BLOCK_NUMBER: u64 = 7280000;
+pub const BYZANTIUM_BLOCK_NUMBER: u64 = 4370000;
+pub const HOMESTEAD_BLOCK_NUMBER: u64 = 1150000;
+
+// Tests can be written to be generate from this file.
+// Tests data will be in a yaml key value format
```
Comment (njgheorghita) [inline simulators/portal-interop/src/constants.rs:12]:
```suggestion
// Test data is written in a yaml key/value format
```

File: simulators/portal-interop/src/constants.rs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+// Tests consist of content key value pairs.
```
Comment (njgheorghita) [inline simulators/portal-interop/src/constants.rs:13]:
```suggestion
// Tests consist of content key/value pairs.
```

File: simulators/portal-interop/src/constants.rs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+// block receipt 1
+// block header 2
+// block receipt 2
+// block header 3
+// would be a valid test but
+// block header 2
+// block receipt 3
+// block header 3
+// this wouldn't be a valid test file as block receipt 3 doesn't have a header place before it
+// related header, block body, receipts must be sequential
```
Comment (njgheorghita) [inline simulators/portal-interop/src/constants.rs:27]: Imo it's not necessary to have this fleshed out example.

File: simulators/portal-interop/src/constants.rs [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+pub const SHANGHAI_BLOCK_NUMBER: u64 = 17034870;
+pub const MERGE_BLOCK_NUMBER: u64 = 15537394;
+pub const LONDON_BLOCK_NUMBER: u64 = 12965000;
+pub const BERLIN_BLOCK_NUMBER: u64 = 12244000;
+pub const ISTANBUL_BLOCK_NUMBER: u64 = 9069000;
+pub const CONSTANTINOPLE_BLOCK_NUMBER: u64 = 7280000;
+pub const BYZANTIUM_BLOCK_NUMBER: u64 = 4370000;
+pub const HOMESTEAD_BLOCK_NUMBER: u64 = 1150000;
+
+// Tests can be written to be generate from this file.
```
Comment (njgheorghita) [inline simulators/portal-interop/src/constants.rs:11]:
```suggestion
// Tests will be automatically generated from the data in this file.
```

File: simulators/portal-interop/src/main.rs [hunk 14]
```diff
@@ -54,155 +64,148 @@ async fn run_suite(host: Simulation, suite: Suite) {
...
 }
 
+fn content_pair_to_string_pair(
+    content_pair: (HistoryContentKey, HistoryContentValue),
+) -> (String, String) {
+    let (content_key, content_value) = content_pair;
+    (content_key.to_hex(), hex_encode(content_value.encode()))
+}
+
+struct ProcessedContent {
```
Comment (njgheorghita) [inline simulators/portal-interop/src/main.rs:74]: Add a docstring

File: simulators/portal-interop/src/main.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (njgheorghita) [reply simulators/portal-interop/src/main.rs:245]: I don't think so? Unless I'm missing something in this diff. I was suggesting that this `NClientTestSpec` should be moved to the `for ProcessedContent ...` loop above (lines 172-208ish)

[the remark this answers is not in the document]
Comment (KolbyML) [reply simulators/portal-interop/src/main.rs:245]: You are thinking of the other NClientTestSpec which I did move there.

image

This one shouldn't be in that for loop

File: simulators/portal-interop/src/constants.rs (hunk 9 above)
Comment (KolbyML) [reply simulators/portal-interop/src/constants.rs:14]: resolved

File: simulators/portal-interop/src/constants.rs (hunk 10 above)
Comment (KolbyML) [reply simulators/portal-interop/src/constants.rs:12]: resolved

File: simulators/portal-interop/src/constants.rs (hunk 12 above)
Comment (KolbyML) [reply simulators/portal-interop/src/constants.rs:27]: resolved

File: simulators/portal-interop/src/constants.rs (hunk 13 above)
Comment (KolbyML) [reply simulators/portal-interop/src/constants.rs:11]: resolved

**Comment by KolbyML** [comment]
image
Answer to this question below
https://github.com/ethereum/portal-hive/pull/114/files#r1378952006

TLDR this is a different test case which shouldn't be in the for loop

File: simulators/portal-interop/src/constants.rs (hunk 8 above)
Comment (KolbyML) [reply simulators/portal-interop/src/constants.rs:29]: > I'm guessing we will eventually add more test data to run these tests for other subnetworks. In which case, I'd imagine we'll have a separate test-file with that data (since it could need special treatment in how it's gossiped). Basically, I think a filename of `./test-data/history.yaml` is sufficient here. The current filename is kinda confusing.

I think that is out of scope for this PR and should be done when support is added to write tests for other networks. I think the format should be ``./test-data/history/nameoftestdata.yaml`` This will match the formatting on the test data repo

> I like the improved description for how the test file should be configured, but I think that it would be better off at the top of the test file itself, since it's not obvious that future devs will see this page when they're adding tests.

There is already a comment in the test data file for this, so we can remove the comment above the constant if you want

> I'd also add a comment listing exactly what tests will be generated & run on each content key/value pair that's added to the test data

I will talk about this in discord dms cause I am confused about it

Comment (njgheorghita) [reply simulators/portal-interop/src/constants.rs:21]: > There is already a comment in the test data file for this

Sure, but from what I can tell it's missing all of the information that I suggested we add....
```
# Test data for Portal Hive
# Data is formatted by comment of block height, header, block body, and receipt, expect for blocks 1 and 100 which don't have receipts
```

So I think it's still worth moving this comment to that test file

**Review by njgheorghita — APPROVED** [review]
:ship:

File: simulators/portal-interop/src/constants.rs (hunk 8 above)
Comment (KolbyML) [reply simulators/portal-interop/src/constants.rs:21]: Oh Ok I think I understand now, I was confused earlier but it makes more sense now

File: simulators/portal-interop/src/constants.rs [hunk 15]
```diff
@@ -8,9 +8,6 @@ pub const CONSTANTINOPLE_BLOCK_NUMBER: u64 = 7280000;
 pub const BYZANTIUM_BLOCK_NUMBER: u64 = 4370000;
 pub const HOMESTEAD_BLOCK_NUMBER: u64 = 1150000;
 
-// Tests will be automatically generated from the data in this file.
-// Test data is written in a yaml key/value format
-// For history tests, the header MUST be placed before its respective block body/receipt, to ensure that the data necessary for validation is available.
 // TESTS THAT USE THIS FILE
```
Comment (njgheorghita) [inline simulators/portal-interop/src/constants.rs:11]: Just to clarify, I think that allll of this comment should be moved over to the test file. Including the list of generated tests

Comment (KolbyML) [reply simulators/portal-interop/src/constants.rs:11]: But the test data file will no longer be located in Portal-Hive in a week

Comment (KolbyML) [reply simulators/portal-interop/src/constants.rs]: Ok I moved the "generated test cases"

Comment (njgheorghita) [reply simulators/portal-interop/src/constants.rs]: Ahhh, interesting. I guess what I'm trying to avoid with these comments is the case where somebody just adds content key/value pairs to the file without understanding the ramifications of doing so (i.e. a set of new tests will be generated & ran in portal-hive). But, the separation of repos does make this tricky. Though, I guess I still think it will be helpful to have these comments inside the new test-data repo. Or maybe moved to a `README` in the new repo...

Comment (KolbyML) [reply simulators/portal-interop/src/constants.rs]: Sounds good
